### PR TITLE
More specific imaging types

### LINKED ISSUE #34: RFC: Better image format/typing
Right now, it's pretty unclear what the shape of an image is, whether it's `(height, width, channels)`, `(channels, height, width)`, or something else. We seem undecided on what it should be, and there's disagreement between the code, the models, and our documentation.

Instead of passing around `ndarray` images, there can be a class that also states the shape of the image and that can be converted to another shape as needed. 

Example usage:
```python
#### imaging_tpyes.py
class Image:
    ...

class ImageDimension(Enum):
    HEIGHT = 'h'
    WIDTH = 'w'
    CHANNELS = 'c'

HEIGHT = ImageDimension.HEIGHT
WIDTH = ImageDimension.WIDTH
CHANNELS = ImageDimension.CHANNELS

### usage
img: Image[HEIGHT, WIDTH, CHANNELS] = Image(img_array)
img_for_other_use = img.change_shape((CHANNELS, HEIGHT, WIDTH))
def some_func(img: Image[CHANNELS, HEIGHT, WIDTH]):
    ...
    
some_func(img) # IDE will complain
some_func(img_for_other_use) # OK
```
This approach shouldn't break anything in runtime when the type hinting is incorrect, but the IDE will give you error squiggles.

@EricPedley @Vince-C156 Let me know what you guys think

### PR DESCRIPTION
**Pull request by MinhxNguyen7**
## Description
Add an Image class that contains an image's data and dimensions order, e.g., `(height, width, channels)` or `(channels, height, width)` along with data validation steps and some manipulation methods. The class also has type subscription to partially ensure that we're not copying arrays/tensors from CPU to GPU willy-nilly.

Example:
```python
image_hwc[np.ndarray] = Image(np.zeros((20, 20, 3)), HWC)
image_chw[torch.Tensor] = Image(torch.zeros((3, 20, 20)), CHW)
image_from_file = Image.from_file("image.png") # defaults to np.uint8 and HWC
```

From now on, any manipulation of image data other than extracting the data to pass it to external code should interface through this class, not the underlying array. Additional interfaces should be added in the future as necessary, e.g., Image.to().

Make the numerical types specified in `imaging_types.py` more specific and fix the code that uses it so that the types check out. This is a bit of a scope creep, so let me know if I should save it for another PR.

- [x] Add tests for Image class
- [x] Add Image class
- [x] Refactor current codebase to use Image class
- [x] More specific number types
- [x] Get rid type errors/type-checker errors in the consumers of these types
- [x] Image load from file

## Testing
I think the new tests cover all of the new code. Everything passes except for the two tests that also fail on main.
image

## Issues
Closes #34 

## Future Follow-up

It's worth considering wrapping YOLO so that we get a clean, type-safe interface for the various modules that use YOLO. For example, calling `self.model.predict(np.repeat(img[...,np.newaxis],3,axis=2))[0].probs.data.numpy()` and like 4 times and ignoring type errors is asking for trouble.

For batch processing, the `Image` class can have a method that outputs an `ImageBatch` class, and the model interfaces can take that.

### CODE REVIEW

**Review by EricPedley — APPROVED** [review]
lgtm
